File: Python/Combination Sum.py
class Solution(object):
    """
    :type candidates: List[int]
    :type target: int
    :rtype: List[List[int]]
    """
    def combinationSum(self, cans, t):
        if cans == None or len(cans) == 0:
            return
        res = []
        cans.sort()
        self.find(cans, t, 0, [], res)
        return res
        
    def find(self, cans, t, index, path, res):
        if t == 0:
            res.append(list(path))
            return
        for i in range(index, len(cans), 1):
            if cans[i] > t:
                break
            path.append(cans[i])
            self.find(cans, t-cans[i], i, path, res)
            del path[len(path)-1]

File: Java/Combination Sum.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] cans, int t) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (cans == null || cans.length == 0)
            return res;
        Arrays.sort(cans);
        find(cans, t, 0, new ArrayList<Integer>(), res);
        return res;
    }
    
    private void find(int[] cans, int t, int index, ArrayList<Integer> path, List<List<Integer>> res) {
        if (t == 0) {
            res.add(new ArrayList<Integer>(path));
            return;
        }
        
        for (int i=index; i<cans.length; i++) {
            if (cans[i] > t) break;
            path.add(cans[i]);
            find(cans, t-cans[i], i, path, res);
            path.remove(path.size()-1);
        }
    }
}